World.py

```diff
@@ -0,0 +1,156 @@
+        - noise2: a boolean indicating whether Noise 2 is present
+        - imitation: a boolean indicating whether players perform imitation
+        - migration: a boolean indicating whether
+        - moore_neighborhood_range: the range of the Moore neighborhood around each cell
+        """
+        self.game = game
+        self.board = board
+        self.players = players
+        self.r = r
+        self.q = q
+        self.noise1 = noise1
+        self.noise2 = noise2
+        self.imitation = imitation
+        self.migration = migration
+        self.moore_neighborhood_range = moore_neighborhood_range
+
+        # randomly insert players in board cells
+        random.shuffle(players)
+        random_cell_sequence = board.random_cell_sequence()
+        for i in range(len(players)):
+            players[i].cell = random_cell_sequence[i]
+            self.board.set_cell(random_cell_sequence[i], players[i])
+            self.board.get_cell(random_cell_sequence[i])
+
+
+    def round(self):
+        """ carries out one round of updates, in which each player is updated
+        once based on the parameters of the world
+        """
+        random.shuffle(self.players)
+        for player in self.players:
+            self.play_with_neighbors(player)
+
+            rand = random.random()
+            if rand < self.r:
+                # randomly reset player strategy and/or location
+                self.noise_update(player)
+
+            else:
+                # perform migration step
+                if self.migration:
+                    self.migration_update(player)
+
+                # perform imitation step
+                if self.imitation:
+                    self.imitation_update(player)
+
+
+    def move_player(self, player, new_cell):
+        """ moves a player from its current cell on the board to the new cell
+        on the board (only if the new cell is unoccupied)
+        """
+        occupied = self.board.get_cell(new_cell) != None
+        if not occupied:
+            self.board.set_cell(player.cell, None)
+            self.board.set_cell(new_cell, player)
+            player.cell = new_cell
+
+
+    def play_with_neighbors(self, player):
+        """ makes a player play a game with its four neighbors and update its
+        total payoff
+        """
+        neighbors = self.board.get_neumann_neighboring_players(player.cell)
+
+        player.payoff = 0
+        for neighbor in neighbors:
+            player.payoff += self.game.play(player, neighbor)[0]
+
+
+    def noise_update(self, player):
+        """ resets the strategy and/or position of the player according to the
+        noise processes Noise 1 and Noise 2 described in the Helbing paper
+        """
+        if self.noise1:
+            # reset strategy
+            rand = random.random()
+            if rand < self.q:
+                player.strategy = Strategy.cooperate
+            else:
+                player.strategy = Strategy.defect
```

Comment:
this function gets called if _rand()<r_ but if noise 1 is active then it tests rand()<q

so for noice 1 we get a probability _r*q_

Reply:
Yep! As I understand the paper, _q_ is the probability that a player resets to "cooperate" _given that_ the player is resetting.